rust/arrow/test/data/mixed_arrays.json

```diff
@@ -1,4 +1,4 @@
-{"a":1, "b":[2.0, 1.3, -6.1], "c":[false, true], "d":4.1}
+{"a":1, "b":[2.0, 1.3, -6.1], "c":[false, true], "d":["4.1"]}
 {"a":-10, "b":[2.0, 1.3, -6.1], "c":null, "d":null}
-{"a":2, "b":[2.0, null, -6.1], "c":[false, null], "d":"text"}
-{"a":3, "b":4, "c": true, "d":[1, false, "array", 2.4]}
+{"a":2, "b":[2.0, null, -6.1], "c":[false, null], "d":["text"]}
+{"a":3, "b":[], "c": [], "d":["array"]}
```

Comment:
@nevi-me the I don't have the full context here, so my change of logic here might not be correct. Basically the current implementation throws out an type incompatible error when a column is detected to have both list and non-list value, for example: `"c": true"` v.s. `"c": [true, false]`.

Do you remember what's the reason for doing automatic conversion between scalar column to scalar list column? This behavior seems a bit too magical to me.

Reply:
There aim was to reduce the odds of throwing away data, converting the scalar to a single-item list.

I can't remember if I checked against Spark, but I'll do that during the day. I'm convinced that I've come across the behaviour elsewhere before, but can't remember where.

Reply:
What have you decided here; to preserve the old behaviour or change it?

Reply:
@nevi-me i just tested it with spark, looks like it's doing the conversion the other way, i.e. when a column contains both scalar and list values, it gets converted to string type. as a result, boolean lists are parsed into `"[true, false]"` string.

I don't have a strong opinion on this, but if we want to match spark, then we should probably go that fallback to string if incompatible types are detected. What do you think? @nevi-me @jorgecarleitao @andygrove .

Reply:
In general I think matching Spark's behavior is probably a good idea so as to try and maximize compatibility with the rest of the ecosystem.

That being said, if Spark's behavior is hard to use, confusing or unwieldy to implement, I think there is a reasonable case to be made for deviating from its behavior too. 

"it depends" lol

Reply:
@houqp @alamb one downside for now is that we don't have utilities to deal with JSON data once it's inferred and interpreted as strings (matching Spark behaviour). For example, we don't have anything to convert "[true, false]" back to an array of those values.

I suppose with schema inference we'll always have to make some compromises. I still prefer casting to list as that loses the least data, but if we use different behaviour, we could document the decision on the relevant function, so that other users could make changes in future (if they motivate for some other behaviour).

Reply:
I have updated the PR to implement the old behavior so the inference rules are consistent with rules used in the actual value conversion code.

After thinking more on this, I think Spark's behavior would be a better design in the long run for the following two reasons:

* It preserves more info from the original JSON data. Once we convert scalar to array of scalar in Arrow, we will lose the information that some of the original values were just scalars. It's better to let users decide how they want to handle incompatible types in their query plans. For example, one might want to simply discard all rows with scalar values as a data cleaning step.
* Converting fields with incompatible types into string is unavoidable for more complex cases anyway. For example, when a field contains scalar, list and object type, there no other option other than reading the value as JSON string.

Perhaps we can refactor the code to match Spark's behavior after we added more JSON parsing kernels.

Reply:
I think that spark's behavior is correct. Our JSON reader and schema inferer is in my opinion very broken for list arrays.

```
{"a": "hello"}
{"a": "world"}
{"a": "this"}
{"a": null}
{"a": "hello"}
{"a": "world"}
{"a": "this"}
{"a": null}
{"a": "hello"}
{"a": "world"}
{"a": "this"}
{"a": null}
{"a": ["a"]}
```

yields a `ListArray<String>` where each scalar is being interpreted as a one-element list. I can't see how this is reasonable.

@houqp I have code to handle arbitrary nested lists on my computer, but it requires that we do not make this kind of crazy stuff, i.e. it works as expected if no casting like the above is allowed.

Reply:
> I have code to handle arbitrary nested lists

@jorgecarleitao could you expand a little bit more on this or share some of the code snippet? Don't we already support reading nested list values?

@nevi-me was mostly concerned with not able to operate on string fields that contains serialized JSON value, I wonder if your code can help alleviate that concern.

Reply:
Sorry, scratch what I said. It is an unrelated issue that I mixed up.

Reply:
Is the conclusion then to keep the existing behaviour? I'm a bit confused 🙃

I went through the code, still a green light from me; so we can merge this

Reply:
Looks like it given Jorge already approved.

Reply:
I do not think we need to/should address it in this PR: this is already a major improvement imo :)